compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -291,167 +358,209 @@ class Semantic {
           if meth.name.toString == "tupled" then Result(value, Nil) // a call like `fun.tupled`
-          else eval(body, thisV, klass, cacheResult = true)
+          else
+            use(env) {
+              eval(body, thisV, klass, cacheResult = true) ++ checkArgs
+            }
 
         case RefSet(refs) =>
-          val resList = refs.map(_.call(meth, superType, source))
+          val resList = refs.map(_.call(meth, args, superType, source))
           val value2 = resList.map(_.value).join
           val errors = resList.flatMap(_.errors)
           Result(value2, errors)
       }
+    }
 
     /** Handle a new expression `new p.C` where `p` is abstracted by `value` */
-    def instantiate(klass: ClassSymbol, ctor: Symbol, source: Tree): Contextual[Result] =
-      value match {
+    def instantiate(klass: ClassSymbol, ctor: Symbol, args: List[ArgInfo], source: Tree): Contextual[Result] = log("instantiating " + klass.show + ", args = " + args, printer, res => res.asInstanceOf[Result].show) {
+      val trace1 = trace.add(source)
+      if promoted.isCurrentObjectPromoted then Result(Hot, Nil)
+      else value match {
         case Hot  =>
-          Result(Hot, Errors.empty)
+          val buffer = new mutable.ArrayBuffer[Error]
+          val args2 = args.map { arg =>
+            val errors = arg.promote
+            buffer ++= errors
+            if errors.isEmpty then Hot
+            else arg.value.widen
+          }
+
+          if buffer.isEmpty then Result(Hot, Errors.empty)
+          else
+            val value = Warm(klass, Hot, ctor, args2)
+            if !heap.contains(value) then
+              val obj = Objekt(klass, fields = mutable.Map.empty, outers = mutable.Map(klass -> Hot))
+              heap.update(value, obj)
+            val res = value.call(ctor, args, superType = NoType, source)
+            Result(value, res.errors)
 
         case Cold =>
-          val error = CallCold(ctor, source, trace.toVector)
+          val error = CallCold(ctor, source, trace1.toVector)
           Result(Hot, error :: Nil)
 
         case addr: Addr =>
+          given Trace = trace1
+
           // widen the outer to finitize addresses
           val outer = addr match
-            case Warm(_, _: Warm) => Cold
+            case Warm(_, _: Warm, _, _) => Cold
             case _ => addr
 
-          val value = Warm(klass, outer)
+          val value = Warm(klass, outer, ctor, args.map(_.value).widen)
           if !heap.contains(value) then
             val obj = Objekt(klass, fields = mutable.Map.empty, outers = mutable.Map(klass -> outer))
             heap.update(value, obj)
-          val res = value.call(ctor, superType = NoType, source)
-          Result(value, res.errors)
+          val res = value.call(ctor, args, superType = NoType, source)
 
-        case Fun(body, thisV, klass) =>
-          ??? // impossible
+          // Abstract instances of local classes inside 2nd constructor as Cold
```

Comment:
How do we know that we're instantiating a local class? Does 2nd constructor mean a constructor that was called from a constructor or does it mean secondary constructor?